software/controller/lib/hal/adc.cpp

```diff
@@ -168,29 +171,48 @@ void HalApi::InitADC() {
 
   adc->adc[0].configuration2.oversampling_ratio = (OversampleLog2 > 0) ? OversampleLog2 - 1 : 0;
 
-  // \TODO: what is 4?
+  // Set oversampling shift if necessary (see [RM] Table 66)
   adc->adc[0].configuration2.oversampling_shift = (OversampleLog2 < 4) ? 0 : (OversampleLog2 - 4);
 
-  // Set sample times ([RM] 16.4.12). I'm using 92.5 A/D clocks to sample.
-  static_assert(AdcSampleTime == 92);
-  // \todo all of these are = 5 -- should this be a single variable based on something above?
-  adc->adc[0].sample_times.ch5 = 5;
-  adc->adc[0].sample_times.ch6 = 5;
-  adc->adc[0].sample_times.ch9 = 5;
-  adc->adc[0].sample_times.ch15 = 5;
-  adc->adc[0].sample_times.ch2 = 5;
-
-  // Set conversion sequence length:
-  // \TODO: why - 1 ?
+  // Set sample times ([RM] 16.4.12).
+  auto ADCSampleTimeRegister = [&](ADCSampleTime sample_time) {
+    switch (sample_time) {
+      case ADCSampleTime::s2_5:
```

Comment:
So, this function makes me wonder if maybe we should implement this Enum the other way around -- have each value of the Enum be fundamentally defined as an `uintSOMETHING_t` (whatever those `ch*` registers take) as the underlying type. And then have a function that maps each one in `ClocksToSample` which then returns something like `92.5` as a float (or truncated int). Then the register assignments below would just  be casts, and the function call would be the one up top where 13 is added?

Reply:
I actually started that way, and didn't like having to cast to `uint` when setting the `adc->adc[0].sample_times.chx` registers, but i think it makes a lot of sense to do it like this! will give it a shot.